### PR TITLE
[DOC-2728] [DOC-2886] - CodeCov tutorial, run step setting, run script in run step

### PR DESCRIPTION
**Pull request by aimurphy**
DOC-2728: Added a tutorial for CodeCov.
- Diff: https://github.com/harness/developer-hub/pull/1177/files#diff-24f9c8c99713131ac11f71b17f8162c4e6ff1af8bfe67819320f3931540c0e55
- Preview: https://642b1277ab8bc450ef3e3db3--harness-developer.netlify.app/tutorials/build-code/ci-tutorial-codecov-test

Refresh Run a script in a Build stage
- Diff: https://github.com/harness/developer-hub/pull/1177/files#diff-7c5af8c0c0b9abfedfa43bc71d4cee7da5e61289aa06826c154d2c12125c456d
- Preview: https://642b1277ab8bc450ef3e3db3--harness-developer.netlify.app/docs/continuous-integration/use-ci/run-ci-scripts/run-a-script-in-a-ci-stage

CI Run step settings - Container Registry/Image field requirements by build infra; Output variable instructions
- Diff: https://github.com/harness/developer-hub/pull/1177/files#diff-265e13dfbbd707a144211ba6cd480b9af9e7197ed53aaa265d9b2a1771d5968e
- Preview: https://642b1277ab8bc450ef3e3db3--harness-developer.netlify.app/docs/continuous-integration/ci-technical-reference/run-step-settings
- DOC-2886

Run Tests step settings - Container Registry/Image field requirements by build infra; Output variable instructions
- Diff: https://github.com/harness/developer-hub/pull/1177/files#diff-650f99386ef822aee9bebaaf19fbb9dfa783234fffb4eb82224a0fe8d7339493
- Preview: https://642b1277ab8bc450ef3e3db3--harness-developer.netlify.app/docs/continuous-integration/ci-technical-reference/configure-run-tests-step-settings
- DOC-2886

Other changes:
- Fixed some delegate links.
- Added a link to the CodeCov tutorial from the CI reference docs.
- Updated titles of a couple topics & the links referencing those titles.

### CODE REVIEW

File: tutorials/build-code/ci-tutorial-codecov-test.md [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+```mdx-code-block
+<Tabs>
+  <TabItem value="Visual" label="Visual" default>
+```
+
+1. In Harness, go to the pipeline where you want to include CodeCov.
+2. Add a [Run step](/docs/continuous-integration/ci-technical-reference/run-step-settings).
+3. Enter a **Name** for the step.
+4. For **Container Registry**, select your preferred Docker connector or use the Harness default Docker connector
+5. For **Image**, enter the Docker image needed to run your tests, such as `python:latest`.
```
Comment (hemanthmantri) [inline tutorials/build-code/ci-tutorial-codecov-test.md:49]: Let's not specify `Container Registry` and `Image`. They are optional and not needed on Cloud Infra.

File: tutorials/build-code/ci-tutorial-codecov-test.md [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+6. In the **Command** field, include all commands necessary to prepare the test environment, run tests with code coverage, declare or ingest the `CODECOV_TOKEN` environment variable, and download and run the CodeCov Uploader tool. For example:
```
Comment (hemanthmantri) [inline tutorials/build-code/ci-tutorial-codecov-test.md:50]: Export `CODECOV_TOKEN` environment variable

File: tutorials/build-code/ci-tutorial-codecov-test.md [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+
+   ```sh
+   echo "Welcome to Harness CI"
+   uname -a
+   pip install pytest
+   pip install pytest-cov
```
Comment (hemanthmantri) [inline tutorials/build-code/ci-tutorial-codecov-test.md:56]: Please check if the above two are present on Harness Cloud VM already. We can remove them.

File: tutorials/build-code/ci-tutorial-codecov-test.md [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+   pip install -r requirements.txt
+
+   pytest -v --cov --junitxml="result.xml" test_api.py test_api_2.py test_api_3.py
+
+   #Declare Codecov Upload Token
+   # ff3e...d6a4
+   CODECOV_TOKEN=3bcc...65623
```
Comment (hemanthmantri) [inline tutorials/build-code/ci-tutorial-codecov-test.md:63]: Remove this. Let's not encourage the usage of keys 🔑  in plain text. Let's only recommend environment variable exporting with a secret.

File: tutorials/build-code/ci-tutorial-codecov-test.md [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+   # Download Codecov uploader and run it.
+   curl -Os https://uploader.codecov.io/latest/linux/codecov
+   chmod +x codecov
+   ./codecov -t ${CODECOV_TOKEN}
+
+   echo "Done"
```
Comment (hemanthmantri) [inline tutorials/build-code/ci-tutorial-codecov-test.md:70]: Not needed

File: tutorials/build-code/ci-tutorial-codecov-test.md [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+   ```
+
+<details>
+<summary>Use a text secret for the CodeCov Upload Token</summary>
+
+For the `CODECOV_TOKEN` environment variable, you can either declare the token directly in the **Command** or store and reference the token as a [text secret](/docs/platform/Security/add-use-text-secrets).
```
Comment (hemanthmantri) [inline tutorials/build-code/ci-tutorial-codecov-test.md:76]: Only secret.

File: tutorials/build-code/ci-tutorial-codecov-test.md [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+
+pytest -v --cov --junitxml="result.xml" test_api.py test_api_2.py test_api_3.py
+
+# Download Codecov uploader and run it.
+curl -Os https://uploader.codecov.io/latest/linux/codecov
+chmod +x codecov
+./codecov -t ${CODECOV_TOKEN}
+
+echo "Done"
+```
```
Comment (hemanthmantri) [inline tutorials/build-code/ci-tutorial-codecov-test.md:100]: Let's have only one version of YAML with secret/env variable.

File: tutorials/build-code/ci-tutorial-codecov-test.md [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+                      ./codecov -t ${CODECOV_TOKEN}
+
+                      echo "Done"
+                    reports:
+                      type: JUnit
+                      spec:
+                        paths:
+                          - "**/*.xml"
+```
+
```
Comment (hemanthmantri) [inline tutorials/build-code/ci-tutorial-codecov-test.md:231]: Let's only have one YAML snippet for this tutorial and add comments to the steps.

File: tutorials/build-code/ci-tutorial-codecov-test.md [hunk 9]
```diff
@@ -0,0 +1,301 @@
+---
+sidebar_position: 140
+description: Add CodeCov code coverage to a CI pipeline.
+keywords: [Hosted Build, Continuous Integration, Hosted, CI Tutorial,Go,Ko,Cosign]
```
[the remark this answers is not in the document]
Comment (aimurphy) [reply tutorials/build-code/ci-tutorial-codecov-test.md:4]: sorry I copied the metadata from another topic and forgot to update it.

File: tutorials/build-code/ci-tutorial-codecov-test.md (hunk 4 above)
Comment (aimurphy) [reply tutorials/build-code/ci-tutorial-codecov-test.md:63]: changed to use secret

**Comment by aimurphy** [comment]
@hemanthmantri Could you please review the tutorial again. I made changes based on your comments.

File: docs/continuous-integration/ci-technical-reference/run-step-settings.md [hunk 10]
```diff
@@ -91,70 +97,78 @@ For more information, go to the [Built-in Harness Variables Reference](../../pla
 
 ### Output Variables
```
Comment (aimurphy) [inline docs/continuous-integration/ci-technical-reference/run-step-settings.md:98]: @jimsheldon 

https://642b1277ab8bc450ef3e3db3--harness-developer.netlify.app/docs/continuous-integration/ci-technical-reference/run-step-settings

File: docs/continuous-integration/ci-technical-reference/configure-run-tests-step-settings.md [hunk 11]
```diff
@@ -96,26 +103,63 @@ Variables passed to the container as environment variables and used in the step'
 
 ### Output Variables
```
[the remark this answers is not in the document]
Comment (jimsheldon) [reply docs/continuous-integration/ci-technical-reference/configure-run-tests-step-settings.md:104]: LGTM!

**Review by jimsheldon — DISMISSED** [review]

File: tutorials/build-code/ci-tutorial-codecov-test.md [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+```
+```mdx-code-block
+<Tabs>
+  <TabItem value="Visual" label="Visual" default>
+```
+
+1. In Harness, go to the **Build** stage in the pipeline where you want to include CodeCov.
+2. Select **Add Step**, select **Add Step** again, and then select **Run** in the Step Library.
```
Comment (douglas-j-bothwell) [inline tutorials/build-code/ci-tutorial-codecov-test.md:49]: I mentioned this in another PR, not sure if was one of yours but anyway: I generally say "click" or "choose" for UI elements that you simply click to go from one place or state to another, like **Add Stage** or a hyperlink vs. "select" for things like radio buttons and checkboxes, where you click something and it has the state of being selected vs. unselected.

File: docs/continuous-integration/ci-technical-reference/run-step-settings.md [hunk 13]
```diff
@@ -24,24 +24,30 @@ Enter a name summarizing the step's purpose. Harness automatically assigns an **
...
-The FQN (fully-qualified name) of the Docker image to use when this step runs commands, for example `us.gcr.io/playground-123/quickstart-image`.
-
-The image name should include the tag. If you don't include a tag, Harness uses the latest tag.
+The **Image** is the FQN (fully-qualified name) or artifact name of the Docker image to use when this step runs commands, for example `us.gcr.io/playground-123/quickstart-image`. The image name should include the tag. If you don't include a tag, Harness uses the latest tag.
 
 You can use any Docker image from any Docker registry, including Docker images from private registries. Different container registries require different name formats:
 
-* **Docker Registry:** Input the name of the artifact you want to deploy, such as `library/tomcat`. Wildcards aren't supported.
-* **GCR:** Input the FQN (fully-qualified name) of the artifact you want to deploy. Images in repos must reference a path, for example: `us.gcr.io/playground-123/quickstart-image:latest`.
+* **Docker Registry:** Input the name of the artifact you want to deploy, such as `library/tomcat`. Wildcards aren't supported. FQN is required for images in private container registries.
```
Comment (douglas-j-bothwell) [inline docs/continuous-integration/ci-technical-reference/run-step-settings.md:35]:
```suggestion
* **Docker Registry:** Input the name of the artifact you want to deploy, such as `library/tomcat`. Wildcards aren't supported. FQN is required for images in private container registries.
```
"Wildcards aren't supported" -- Says this for Docker Registry but not the other two, which kinda implies that maybe they _do_ support wildcards. Which might be true, or not. Perhaps clarify?

File: docs/continuous-integration/use-ci/run-ci-scripts/run-a-script-in-a-ci-stage.md [hunk 14]
```diff
@@ -1,82 +1,152 @@
...
+                  type: Run
+                  name: Run pytest # Specify a name for the step.
+                  identifier: Run_pytest # Define a step ID, usually based on the name.
+                  spec:
+                    shell: Sh
+                    command: |-
+                      # Enter commands/script.
+```
+
+Depending on your build infrastructure and the commands you are running, `connectorRef` and `image` may be required. These settings define a container registry connector and image containing the binaries that the step needs to run your script. For example, a cURL scripts may require a cURL image, such as `curlimages/curl:7.73.0`. For information about when these settings are required and how to specify images, go to the [Run step settings reference](../../ci-technical-reference/run-step-settings.md). The following example shows a `Run` step with `connectorRef` and `image`:
```
Comment (douglas-j-bothwell) [inline docs/continuous-integration/use-ci/run-ci-scripts/run-a-script-in-a-ci-stage.md:90]:
```suggestion
Depending on your build infrastructure and the commands you are running, `connectorRef` and `image` might be required. These settings define a container registry connector and image containing the binaries that the step needs to run your script. For example, a cURL script might require a cURL image, such as `curlimages/curl:7.73.0`. For information about when these settings are required and how to specify images, go to the [Run step settings reference](../../ci-technical-reference/run-step-settings.md). The following example shows a `Run` step with `connectorRef` and `image`:
```

File: docs/continuous-integration/use-ci/run-ci-scripts/run-a-script-in-a-ci-stage.md [hunk 15]
```diff
@@ -1,82 +1,152 @@
...
+                  identifier: Run_pytest # Define a step ID, usually based on the name.
+                  spec:
+                    connectorRef: account.harnessImage # Specify a container registry, if required.
+                    image: python:latest # Specify an image, if required.
+                    shell: Sh
+                    command: |-
+                      # Enter a command or script.
+```
+
+Define other settings as necessary. For example, if the `command` runs tests, you might specify report paths, and, if your script requires a token or secret, you might supply the token as `envVariables` For information about these settings, go to the [Run step settings reference](../../ci-technical-reference/run-step-settings.md). The following example includes `reports` settings and commands that run tests.
```
Comment (douglas-j-bothwell) [inline docs/continuous-integration/use-ci/run-ci-scripts/run-a-script-in-a-ci-stage.md:105]:
```suggestion
Define other settings as necessary. For example, if the `command` runs tests, you might specify report paths.  If your script requires a token or secret, you might supply the token as `envVariables`. For information about these settings, go to the [Run step settings reference](../../ci-technical-reference/run-step-settings.md). 

The following example includes `reports` settings and commands that run tests.
```

File: docs/continuous-integration/ci-technical-reference/run-step-settings.md [hunk 16]
```diff
@@ -91,70 +97,78 @@ For more information, go to the [Built-in Harness Variables Reference](../../pla
...
-<+pipeline.stages.[stage name].spec.execution.steps.[step name].output.outputVariables.[varName]>
+<+stages.[stageID].spec.execution.steps.[stepID].output.outputVariables.[varName]>
+<+pipeline.stages.[stageID].spec.execution.steps.[stepID].output.outputVariables.[varName]>
 ```
 
-You can also access environment variables through the auto-suggest/ auto-complete feature in the Harness UI.
+<details>
+<summary>YAML example: Output variable</summary>
+
+In the following YAML example, step `a` exports an output variable called `myVar`, and then step `b` references that output variable.
```
Comment (douglas-j-bothwell) [inline docs/continuous-integration/ci-technical-reference/run-step-settings.md:135]:
```suggestion
In the following YAML example, step `a` exports an output variable called `myVar`, and then step `b` references that output variable.
```
Consider using `alpha` and `beta` instead. These one-letter variables are kinda of easy to read over.

File: docs/continuous-integration/use-ci/run-ci-scripts/run-a-script-in-a-ci-stage.md (hunk 14 above)
Comment (douglas-j-bothwell) [inline docs/continuous-integration/use-ci/run-ci-scripts/run-a-script-in-a-ci-stage.md:90]:
```suggestion
Depending on your build infrastructure and the commands you are running, `connectorRef` and `image` might be required. These settings define a container registry connector and image containing the binaries that the step needs to run your script. For example, a cURL script may require a cURL image such as `curlimages/curl:7.73.0`. For information about when these settings are required and how to specify images, go to the [Run step settings reference](../../ci-technical-reference/run-step-settings.md). 

The following example shows a `Run` step with `connectorRef` and `image`:
```

**Review by douglas-j-bothwell — APPROVED** [review]
A few tweaks/sugs but otherwise looks good!

**Pull request merged**